File: Leetcode 1501-1600/1536/1536.py
# Leetcode 1536: Minimum Swaps to Arrange a Binary Grid
# https://leetcode.com/problems/minimum-swaps-to-arrange-a-binary-grid/
# Solved on 2nd of March, 2026
class Solution:
    def minSwaps(self, grid: list[list[int]]) -> int:
        """
        Calculates the minimum number of adjacent row swaps needed to make the grid valid.
        A grid is valid if all cells above the main diagonal are zeros.

        :param grid: A square n x n binary grid.
        :return: The minimum swaps required, or -1 if it's impossible.
        """
        gridSize = len(grid)
        trailingZeros = []
        for row in grid:
            zeroCount = 0
            for i in range(gridSize - 1, -1, -1):
                if row[i] == 0:
                    zeroCount += 1
                else:
                    break
            trailingZeros.append(zeroCount)

        totalSwaps = 0
        for i in range(gridSize):
            targetZeros = gridSize - 1 - i
            foundIndex = -1
            for j in range(i, gridSize):
                if trailingZeros[j] >= targetZeros:
                    foundIndex = j
                    break

            if foundIndex == -1:
                return -1

            totalSwaps += foundIndex - i
            poppedValue = trailingZeros.pop(foundIndex)
            trailingZeros.insert(i, poppedValue)

        return totalSwaps

File: Leetcode 1501-1600/1536/1536.java
// Leetcode 1536: Minimum Swaps to Arrange a Binary Grid
// https://leetcode.com/problems/minimum-swaps-to-arrange-a-binary-grid/
// Solved on 2nd of March, 2026
class Solution {
    /**
     * Calculates the minimum number of adjacent row swaps to arrange a binary grid
     * such that all elements above the main diagonal are zero.
     * @param grid An n x n binary grid.
     * @return The minimum swaps required, or -1 if it is impossible.
     */
    public int minSwaps(int[][] grid) {
        int n = grid.length;
        int[] trailingZeros = new int[n];
        
        for (int i = 0; i < n; i++) {
            int count = 0;
            for (int j = n - 1; j >= 0; j--) {
                if (grid[i][j] == 0) {
                    count++;
                } else {
                    break;
                }
            }
            trailingZeros[i] = count;
        }
        
        int totalSwaps = 0;
        
        for (int i = 0; i < n; i++) {
            int targetZeros = n - 1 - i;
            int currentRow = i;
            
            while (currentRow < n && trailingZeros[currentRow] < targetZeros) {
                currentRow++;
            }
            
            if (currentRow == n) {
                return -1;
            }
            
            totalSwaps += currentRow - i;
            int temp = trailingZeros[currentRow];
            
            for (int k = currentRow; k > i; k--) {
                trailingZeros[k] = trailingZeros[k - 1];
            }
            trailingZeros[i] = temp;
        }
        
        return totalSwaps;
    }
}